lib/cosmos/dart/script/dart_reducer.rb

```diff
@@ -0,0 +1,277 @@
+                    end
+                  end
+                  avg_value = avg_value / total_samples if total_samples != 0
+                  max_value = Float::NAN if max_nan_found and !max_value
+                  min_value = Float::NAN if min_nan_found and !min_value
+                  avg_value = Float::NAN if avg_nan_found and total_samples == 0
+                  new_row.write_attribute(max_item_name, max_value)
+                  new_row.write_attribute(min_item_name, min_value)
+                  new_row.write_attribute(avg_item_name, avg_value)
+                end
+                # Need to transaction the below as it is possible to create reduced and not update original
+                new_row.save!
+                Cosmos::Logger.info("Created #{new_row.class}:#{new_row.id} with #{mappings.length} items from #{new_row.num_samples} samples")
+                model1.where(id: sample_rows.map(&:id)).update_all(:reduced_state => 2)
+                rows = rows[-1..-1]
+              end
+            end
+          else
+            Cosmos::Logger.debug("No Mappings for JobType #{job_type}: #{packet_config_id}: #{table_index}")
+          end
+        rescue Exception => error
+          Cosmos::Logger.error("Reducer Thread Exception: #{error.formatted}")
+        end
+        Cosmos::Logger.info("Completed JobType #{job_type}: #{packet_config_id}: #{table_index}")
+
+        @mutex.synchronize do
+          @locked_tables.delete([job_type, packet_config_id, table_index])
+        end
+        @master_queue << @thread_queue
+      end
+    rescue Exception => err
+      Cosmos::Logger.error("Reducer Thread Unexpectedly Died: #{err.formatted}")
+    end
+
+    Cosmos::Logger.info("Reducer Thread #{@instance_num} Shutdown")
+  end
+
+  def shutdown
+    @running = false
+    @thread_queue << nil
+  end
+
+  def join
+    Cosmos.kill_thread(self, @thread)
+  end
+
+  def graceful_kill
+    shutdown()
+  end
+end
+
+class DartReducer
+  include DartCommon
+
+  def initialize(num_threads = 5)
+    Cosmos::Logger.info("Dart Reducer Starting with #{num_threads} threads...")
+    @master_queue = Queue.new
+    @locked_tables = []
+    @mutex = Mutex.new
+    setup_threads(num_threads)
+  end
+
+  def setup_threads(num_threads)
+    @threads = []
+    num_threads.times do |index|
+      @threads << DartReducerWorkerThread.new(@master_queue, @locked_tables, @mutex, index + 1)
+    end
+  end
+
+  def shutdown
+    @threads.each {|thread| thread.shutdown}
+    @threads.each {|thread| thread.join}
+  end
+
+  def get_table_model(table, reduction_modifier = "")
+    model_name = "T" + table[1..-1] + reduction_modifier
+    begin
+      model = Cosmos.const_get(model_name)
+    rescue
+      # Need to create model
+      model = Class.new(ActiveRecord::Base) do
+        self.table_name = table + reduction_modifier
+      end
+      Cosmos.const_set(model_name, model)
+    end
+    return model
+  end
+
+  def start
+    begin
+      while true
```

Comment:
Doesn't this code continuously find the same tables each cycle and send them for processing? Then the worker thread has to do some work to determine that no work needs to be done and we go around again. Seems like there's a bunch of overhead processing the same things each cycle which don't need to be processed anymore.

Reply:
It checks all the tables every minute.   Could be optimized to only check the hour tables and the day tables on the hour/day transition (and probably always on startup).